bookshelf/test/ControllersTest.php

```diff
@@ -0,0 +1,174 @@
+            'image' => new UploadedFile(
+                __DIR__ . '/../images/moby-dick.png',
+                'moby-dick.png',
+                'image/png',
+                null,
+                true
+            ),
+        ]);
+        $this->assertEquals(302, $response->getStatusCode());
+        $crawler = new Crawler($response->getContent());
+        $redirectUri = $crawler->filter('a')->attr('href');
+        $response = $this->call('GET', $redirectUri);
+
+        $this->assertContains(
+            'moby-dick.png',
+            $response->getContent()
+        );
+    }
+
+    /**
+     * @depends testCreateBook
+     */
+    public function testPagination()
+    {
+        // Create another book
+        $response = $this->call('POST', '/books/add', [
+            'title' => 'Treasure Island',
+            'author' => 'Robert Louis Stevenson',
+            'published_date' => '1883-01-01',
+            'description' => '',
+        ]);
+        $this->assertEquals(302, $response->getStatusCode());
+
+        // Now go through the pages one by one and confirm we have the books
+        $response = $this->call('GET', '/');
+        $this->assertEquals(200, $response->getStatusCode());
+        $crawler = new Crawler($response->getContent());
+
+        $this->assertEquals(1, $crawler
+            ->filter('h4:contains("The Cat in the Hat")')->count());
+        $more = $crawler->filter('a:contains("More")');
+        $this->assertEquals(1, $more->count());
+
+        $response = $this->call('GET', $more->attr('href'));
+        $this->assertEquals(200, $response->getStatusCode());
+        $crawler = new Crawler($response->getContent());
+
+        $this->assertEquals(1, $crawler
+            ->filter('h4:contains("Treasure Island")')->count());
+    }
+
+    /**
+     * @depends testCreateBook
+     */
+    public function testEditBook()
+    {
+        $response = $this->call('GET', '/');
+        $this->assertEquals(200, $response->getStatusCode());
+        $crawler = new Crawler($response->getContent());
+
+        // Find the first book and get the edit URL
+        $firstBook = $crawler->filter('a:contains("The Cat in the Hat")');
+        $this->assertEquals(1, $firstBook->count());
+
+        // Edit the book
+        $response = $this->call('POST', $firstBook->attr('href') . '/edit', [
+            'title' => 'The Cat in the Hat (edited)',
+            'description' => '**New Description**'
+        ]);
+
+        $this->assertEquals(302, $response->getStatusCode());
+        $crawler = new Crawler($response->getContent());
+        $redirectUri = $crawler->filter('a')->attr('href');
+        $response = $this->call('GET', $redirectUri);
+
+        $this->assertContains(
+            'The Cat in the Hat (edited)',
+            $response->getContent()
+        );
+        $this->assertContains('**New Description**', $response->getContent());
+    }
+
+    public function testLogging()
+    {
+        // TODO
```

Comment:
Friendly reminder on this guy :)

Reply:
Done! Added both functional and E2E tests!